File: SupportDemo/app/src/main/java/cn/six/sup/rv/swipe_menu/SwipeMenuLayout.java
package cn.six.sup.rv.swipe_menu;

import android.content.Context;
import android.support.v4.widget.ViewDragHelper;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;

/**
 * Created by songzhw on 2016-09-08
 */
public class SwipeMenuLayout extends FrameLayout {
    private ViewDragHelper dragger;
    private View contentView, menuView;
    private int menuWidth;

    public SwipeMenuLayout(Context context) {
        super(context);
    }

    public SwipeMenuLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        contentView = getChildAt(0);
        menuView = getChildAt(1);
        dragger = ViewDragHelper.create(this, callback);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        menuWidth = menuView.getMeasuredWidth();
        contentView.layout(left, top, right, bottom);
        menuView.layout(right, top, right + menuWidth, bottom);

        System.out.println("szw layout MenuWidth = " + menuWidth);
    }

    private ViewDragHelper.Callback callback = new ViewDragHelper.Callback() {
        @Override
        public boolean tryCaptureView(View child, int pointerId) {
            return child == contentView;
        }

        @Override
        public int clampViewPositionHorizontal(View child, int left, int dx) {
            if (left < -menuWidth) {
                return -menuWidth;
            } else if (left > 0) {
                return 0;
            }
            return left;
        }

        @Override
        public void onViewPositionChanged(View changedView, int left, int top, int dx, int dy) {
            super.onViewPositionChanged(changedView, left, top, dx, dy);
            System.out.println("szw left = " + left + " ; dx = " + dx);
            menuView.offsetLeftAndRight(dx);
        }
    };

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        return dragger.shouldInterceptTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        dragger.processTouchEvent(event);
        return true;
    }

    @Override
    public void computeScroll() {
        if (dragger.continueSettling(true)) {
            invalidate();
        }
    }
}

/*
两个问题

1. 滑动过半, 应该自动就位

2. (bug) 第一项滑出菜单, 再垂直滑到第二屏, 因为复用的关系 , 第16项也变成滑出了菜单(但我是没有滑的, 只是复用view的关系引起了混乱)
 */

File: ConstraintLayoutDemo/app/src/main/java/ca/six/ctlay/utils/rv/SwipeMenuLayout.kt
package ca.six.advk.utils.rv

import android.content.Context
import android.support.v4.view.ViewCompat
import android.support.v4.widget.ViewDragHelper
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout


class SwipeMenuLayout : FrameLayout {
    private var self: SwipeMenuLayout? = null
    private var dragger: ViewDragHelper? = null
    private var contentView: View? = null
    private var menuView: View? = null
    private var menuWidth: Int = 0
    private var draggedDistance: Int = 0
    var isOpen = false

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}


    override fun onFinishInflate() {
        super.onFinishInflate()
        self = this
        contentView = getChildAt(0)
        menuView = getChildAt(1)
        dragger = ViewDragHelper.create(this, callback)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        menuWidth = menuView!!.measuredWidth
        contentView!!.layout(left, top, right, bottom)
        menuView!!.layout(right, top, right + menuWidth, bottom)
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        return dragger!!.shouldInterceptTouchEvent(ev)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        dragger!!.processTouchEvent(event)
        return true
    }

    override fun computeScroll() {
        if (dragger!!.continueSettling(true)) {
            invalidate()
        }
    }


    private val callback = object : ViewDragHelper.Callback() {
        override fun tryCaptureView(child: View, pointerId: Int): Boolean {
            return child === contentView
        }

        override fun clampViewPositionHorizontal(child: View, left: Int, dx: Int): Int {
            if (left < -menuWidth) {
                return -menuWidth
            } else if (left > 0) {
                return 0
            }
            return left
        }

        override fun onViewPositionChanged(changedView: View, left: Int, top: Int, dx: Int, dy: Int) {
            super.onViewPositionChanged(changedView, left, top, dx, dy)
            draggedDistance = left // menuWidth是246. 左滑拉出菜单时, left从0一直变到-246.  反之则是一直从负数变成0
            menuView!!.offsetLeftAndRight(dx)
        }

        override fun onViewReleased(releasedChild: View, xvel: Float, yvel: Float) { //后两参是velocity
            if (releasedChild !== contentView) {
                return   //对menuView不要滑, 只要点击即可
            }

            val distance = Math.abs(draggedDistance) //大于0, 说明菜单拉出来了一部分.  若为0, 才表示菜单没出来.
            val threshold = menuWidth / 2
            if (distance > threshold) { //拉出了一半多, 这时松手, 要回到拉出的状态
                open()
            } else {
                close()
            }

        }
    }

    fun open() {
        isOpen = true
        if (dragger!!.smoothSlideViewTo(contentView!!, -menuWidth, 0)) {
            ViewCompat.postInvalidateOnAnimation(self)
        }
    }

    fun close() {
        isOpen = false
        if (dragger!!.smoothSlideViewTo(contentView!!, 0, 0)) {
            ViewCompat.postInvalidateOnAnimation(self)
        }
    }

}